lib/tests/formulae_dependents.rb

```diff
@@ -166,6 +166,23 @@ def install_dependent(dependent, testable_dependents, args:, build_from_source:
           test "brew", "link", dependent.full_name
         end
         test "brew", "install", "--only-dependencies", dependent.full_name
+
+        dependent.deps.each do |dependency|
+          dependency_f = dependency.to_formula
+
+          # Uninstall build- or test-only dependencies to avoid `brew linkage --test`
+          # mistakenly succeeding. Homebrew/homebrew-test-bot#611. Link them otherwise
+          # to avoid spurious CI failures.
+          subcommand = if (dependency.build? || dependency.test?) && dependency_f.any_version_installed?
+            ["uninstall", "--ignore-dependencies"]
```

Comment:
Why do we ignore dependencies here, does this not mean we risk uninstalling stuff actually required?

Reply:
We ignore dependencies because something else installed might need the build/test-only dependency we want to uninstall.

One thing I need to think more about here is the case where the build/test-only dependency is a modified formula: I suspect uninstalling those will require special reinstall logic which we currently don't have.

Reply:
> We ignore dependencies because something else installed might need the build/test-only dependency we want to uninstall.

Do we install it again later if it's needed?

> One thing I need to think more about here is the case where the build/test-only dependency is a modified formula: I suspect uninstalling those will require special reinstall logic which we currently don't have.

I think we have that logic just not for dependants (uses the bottles) or we just don't install them in that case. The existing logic I linked in the issue should help with these edge cases.